### ISSUE TITLE
PythonEvalUTest reinitialises embedded Python interpreter in an unsupported way

### ISSUE THREAD

**Issue by ferrouswheel** [opencog/atomspace#2213]
This is essentially a continuation of #1888 - but opening a new issue as suggested by the last comment there.

The issue `no attribute __reduce_cython__` appears to still happen with Python 3.5/3.6 and cython 0.26 when running PythonEvalUTest. This causes the test to fail.

The last comment on https://github.com/cython/cython/issues/1953 suggests upgrading to cython 0.28 may get something working, although also suggests resetting the interpreter in this way with extension modules is not actually supported by Python.

Sticking with Python version 3.5 to reduce variables involved, I upgraded to cython 0.28.6 which results in a segfault in `PythonEvalUTest` and `PythonUtilitiesUTest`.

Upgrading to the latest cython 0.29.10, results in:
```
[2019-06-09 23:57:49:728] [ERROR] Python error: Interpreter change detected - this module can only be loaded into one interpreter per process.. (/home/joel/work/opencog/atomspace/opencog/cython/PythonEval.cc:694)
```

and subsequent errors.

It feels like we are skirting unsupported territory by re-initialising the python interpreter multiple times in a process, and just got lucky with cython 0.23 and Python 3.5.

**Comment by ferrouswheel** [comment]
Tested Cython 0.23 and Python 3.6.7 - this combination works and passes tests.

For anyone else wanting to test version compatibility, my work flow using [virtualenvwrapper](https://virtualenvwrapper.readthedocs.io/en/latest/) is:

```
mkvirtualenv opencog --python=/usr/bin/python3.5 # or any other version you have installed
# the above puts you in the virtualenv automatically,
# to enable it later, use "workon opencog"
# You can change the version spec to get a particular minor version, this will get the highest
# version in the 0.23.* series
pip install --force cython>=0.23,<0.24
cd atomspace/build
# make sure to forget the old cmake vars related to python
cmake -U CYTHON_EXECUTABLE -U PYTHON_* ..
# i found the change of cython executable wasn't detected, so need to clean the build
make clean
make && make test
deactivate
rmvirtualenv opencog
```

This avoids messing with your system environment, just don't `make install`, or things might get weird.

**Comment by AmeBel** [comment]
opencog/opencog-deps has been update to ubuntu:18.04 with cython version 0.23

**Comment by ferrouswheel** [comment]
@AmeBel Unless you've also implemented something to fix PythonEvalUTest, your PR will break the CI checks.

Since I just spent a tonne of time helping get the test suite working again, part of which was due to you prematurely merging opencog/opencog#3503 - it'd be appreciated if you could try and help us all keep our CI builds green...

**Comment by ferrouswheel** [comment]
Sorry. I reread this issue, and I realise it makes it sound like things are working okay with Ubunutu 18.04 and cython 0.23.

Unfortunately, Ubuntu 18.04 recently upgraded to Python 3.6.8 (from 3.6.7), and this causes PythonEvalUTest to fail again. So while an opencog-deps image may work temporarily. We should prepare for it to fail some time soon.

**Comment by ferrouswheel** [comment]
uhhh actually I have no idea what I'm talking about! :sweat_smile: 

PythonEvalUTest does fail with 3.6.8 for me, but it looks like it's unrelated to the interpretor being reinitialised multiple times.

**Comment by AmeBel** [comment]
The tests are on moses and asmoses are failing but opencog and atomspace are passing on ci, see https://circleci.com/workflow-run/17e9bffc-d7fa-421e-993a-d79bee8d9ae9

**Comment by vsbogd** [comment]
Should we close it after #2361 ?

**Comment by linas** [comment]
Well, see #2411 -- you still have a publicly accessible `initialize` and `finialize` in the API.  You really should get rid of those, first. :-)

**Comment by vsbogd** [comment]
Ah ok, I thought `PythonEvalUTest` doesn't reinitializes Python after #2361 but it still does. So let't keep it open ;-)

**Comment by linas** [comment]
using cython version 0.23 breaks python3, see issue #2550

**Comment by ntoxeg** [comment]
I use Ubuntu 20.04 with Python 3.8 and Cython 0.29.16 and `PythonEvalUTest` passes for me. I actually made a PR moving ocpkg and Dockerfiles to Ubuntu 20.04 but I did that against SingularityNET's repos... I guess that was a mistake?

**Comment by vsbogd** [comment]
`singnet` duplicates `opencog` repos and contains changes which are not in `opencog`. I merge  latest `opencog` changes into `singnet` periodically but it is rare event. I would not recommend mix `opencog` and `singnet` versions because they can be incompatible.

In other words if you made atomspace change in `opencog/atomspace` it would be logical to change `opencog/ocpkg` and `opencog/docker` to have all changes in the same place.

**Comment by vsbogd** [comment]
Moving `ocpkg` to `Ubuntu 20.04` is questionable thing. This script is used by people who would like to try `opencog` locally and requiring latest version of `Ubuntu` seems to be too hard restriction. 

@linas what do you think?

**Comment by ntoxeg** [comment]
Ubuntu 20.04 is the current Long Term Support release, you will have to move on to it sooner or later. I guess I assume everybody uses Docker these days for checking this out, for people using Ubuntu as their OS I think even conservative users will migrate to 20.04.

**Comment by linas** [comment]
> even conservative users will migrate to 20.04.

You are making a false assumption that migration is easy or that its even possible. Some of the users are large users with large deployments, and have no practical way of migrating to new releases. I saw this in hanson robotics: there were many dozens of computers, many of them embedded NUC's, and only one person part-time installing and maintaining. An update to new versions of ubuntu would have required an update to newer versions for ROS .. and there were dozens of github repos with older ROS code that would need to be changed/ported updated. After an update, various devices e.g. video cameras, position tracking sensors, microphones may or may not work (not an ubuntu problem, but a ROS problem) and there were a dozen different types of cameras in use (some experimental, to save money or get higher quality or larger field-of-view or smaller power consumption, higher bandwidth, etc.)  All these require testing and debugging. So a "migration" takes a HUGE amount of effort, and years to accomplish. 

This is the reality that all institutional users face. Soon as you have more than a half-dozen machines, more than just 3 or 4 users, migration is hard, and things break, and have to be manually updated.

**Comment by linas** [comment]
p.s. docker sucks. LXC is wildly superior. (well, I guess its now LXD or LXE...)

**Comment by ntoxeg** [comment]
But we were talking about users "trying it out", not about institutions - `ocpkg` probably sucks for them anyway, it doesn't even manage Python through `pyenv` or `virtualenv`, which is a better practice than polluting your system by invoking `sudo pip`.

Anyway, sooner or later some important components will have to be upgraded for one reason or another. I guess somebody can test if Cython 0.29.16 and Python 3.6 are also going to work, if not then I am afraid I will just have to maintain my own fork and drop this PR.

**Comment by linas** [comment]
> I am afraid I will just have to 

what's the problem? ocpkg has been updated (https://github.com/opencog/ocpkg/issues/148 and https://github.com/opencog/ocpkg/pull/149) and the unit tests for the atomspace now pass. The downstream components are now broken, because various pxd files have gone missing. 

Anyway, I'm unclear as to what all the grumbling is about. Apparently, cython-0.23 was broken. Great. So we're not going to use it any more. That's unrelated to this issue, which is about calling python finalize (which also used to work, but then it broke .. maybe now it works again, with the newer cython?)  

Anyway, I'm closing this issue ... unit tests have been passing for a year.  I don't know if that is because cython got fixed, or if someone disabled the finalize test.  The conversation here is too muddled to keep this issue open; if there are new issues, please open new issues ...

**Comment by AmeBel** [comment]
Previously cython was pegged to version 0.23.5, so as to fix the issue reported here. After change to the latest cython version 0.29.17, by https://github.com/opencog/ocpkg/pull/149, all tests [pass](https://circleci.com/workflow-run/568e7b4d-5d2a-496c-8954-6f856be20a3e.)

Python version used for the current passing test = 3.6.9
